### PR TITLE
chore: split controllers to separate packages + cover them with unit tests

### LINKED ISSUE #393: Move controllers into their own packages
# Short abstract

Right now, all controllers (FeatureFlagConfigurationController, FlagSourceConfigurationController) are located in a common controller package. With this issue I would like to propose to move each of those controllers into their own package.

# Why 

In general, packages should be kept as small as possible to make them easy to understand. Also, in the ginkgo test suite, the controllers might interfere with each other when new tests are added. We recently observed that behaviour in https://github.com/keptn/lifecycle-toolkit, where the tests were failing randomly. Moving the controllers to their own packages will help to isolate them from each other when running the test suites

# Tasks

- [ ] Create a new package for each controller, i.e. `controllers/featureflagconfiguration/controller.go` and `controllers/flagsourceconfiguration/controller.go`
- [ ] Code that is shared between the existing controllers should be moved into a common package, e.g. `controllers/common/common.go`

### PR DESCRIPTION
**Pull request by odubajDT**
## This PR

- moves controllers to separate packages
- moves common functions/constants into separate common package
- migrates ginkgo controller tests to basic unit tests

### Related Issues

Fixes #393

### CODE REVIEW

**Comment by AlexsJones** [comment]
Can you tell us why this is a good idea perhaps? I've not touched this code in a few months and the delta seems large.

**Comment by odubajDT** [comment]
This PR is going to be split into 2 due to the size

**Comment by odubajDT** [comment]
Split of the PR completed, the follow-up PR with additional refactoring and unit tests will follow up

File: controllers/core/featureflagconfiguration/controller_test.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+				cm2.OwnerReferences = append(cm2.OwnerReferences, v1alpha1.GetFfReference(ffConfig2))
+				err := r.Client.Update(ctx, cm2)
+				require.Nil(t, err)
+			}
+
+			_, err = r.Reconcile(ctx, req)
+			require.Nil(t, err)
+
+			ffConfig2 := &v1alpha1.FeatureFlagConfiguration{}
```
Comment (skyerus) [inline controllers/core/featureflagconfiguration/controller_test.go:130]: Is there more appropriate var naming than ffConfig2?

Comment (odubajDT) [reply controllers/core/featureflagconfiguration/controller_test.go]: fixed

**Review by skyerus — APPROVED** [review]
I like the splitting of the controllers package, nice

**Comment by odubajDT** [comment]
> I like the splitting of the controllers package, nice

Thank you!

**Comment by odubajDT** [comment]
> I'll give this a thorough review tomorrow!

Thank you!

File: apis/core/v1alpha1/featureflagconfiguration_types.go [hunk 2]
```diff
@@ -145,3 +145,10 @@ func FeatureFlagConfigurationId(namespace, name string) string {
 func FeatureFlagConfigurationConfigMapKey(namespace, name string) string {
 	return fmt.Sprintf("%s.flagd.json", FeatureFlagConfigurationId(namespace, name))
 }
+
+func (p *FeatureFlagServiceProvider) IsSet() bool {
+	if p == nil || p.Name == "" {
+		return false
+	}
+	return true
```
Comment (toddbaert) [inline apis/core/v1alpha1/featureflagconfiguration_types.go:153]:
```suggestion
	return p != null && p != ""
```

File: controllers/core/flagsourceconfiguration/controller_test.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+			_, err = r.Reconcile(ctx, req)
+			require.Nil(t, err)
+
+			if tt.deployment != nil {
+				deployment := &appsv1.Deployment{}
+				err = fakeClient.Get(ctx, types.NamespacedName{Name: deploymentName, Namespace: testNamespace}, deployment)
+				require.Nil(t, err)
+
+				require.Equal(t, tt.restarted2, deployment.Spec.Template.ObjectMeta.Annotations["kubectl.kubernetes.io/restartedAt"])
```
Comment (toddbaert) [inline controllers/core/flagsourceconfiguration/controller_test.go:102]: Is there a chance that this assertion fails if things are really slow or these tests are reordered? Might it be better to instead test that `kubectl.kubernetes.io/restartedAt` is greater than `restarted2`? Maybe I'm confused abou twhat this test is doing.

**Review by toddbaert — APPROVED** [review]
Approved with some minor comments, questions.

File: controllers/core/flagsourceconfiguration/controller_test.go (hunk 3 above)
Comment (odubajDT) [reply controllers/core/flagsourceconfiguration/controller_test.go:102]: The tests can be reordered and everything should work as expected. Each Iteration is completely independent from the other as we are using `t.Run()`, which is basically a new test with clean context.

 If the test is slow, hard to say to be honest, maybe worth merging and if we have some problems or flakyness in the future, we can always adapt it. Would this work ?

Comment (odubajDT) [reply controllers/core/flagsourceconfiguration/controller_test.go:102]: sure, would better description/comments do the trick for you ?

Comment (Kavindu-Dodan) [reply controllers/core/flagsourceconfiguration/controller_test.go:102]: @odubajDT  IMO yes. May be comment explaining what we do here and better naming for `restarted1` & `restarted2`

Comment (odubajDT) [reply controllers/core/flagsourceconfiguration/controller_test.go]: fixed :+1:

**Review by Kavindu-Dodan — APPROVED** [review]

File: controllers/core/flagsourceconfiguration/controller_test.go (hunk 3 above)
Comment (Kavindu-Dodan) [reply controllers/core/flagsourceconfiguration/controller_test.go]: @toddbaert I think this is fixed. Can you also check ? :)

**Review by james-milligan — APPROVED** [review]

**Pull request merged**